File: Tests/DatabaseTestUtils.m
#import "DatabaseTestUtils.h"\

#import "../RAnalytics/Core/Private/_RAnalyticsDatabase.h"

sqlite3* openRegularConnection() {
    return openConnection(SQLITE_OPEN_READWRITE | SQLITE_OPEN_CREATE);
}

sqlite3* openReadonlyConnection() {
    return openConnection(SQLITE_OPEN_READONLY | SQLITE_OPEN_CREATE);
}

sqlite3* openConnection(int flags) {
    sqlite3* connection;
    
    sqlite3_open_v2("file::memory:?cache=shared", &connection, flags, NULL);
    
    return connection;
}

static _RAnalyticsDatabase* db; // Required because of ARC
_RAnalyticsDatabase* mkDatabase(sqlite3* connection) {
    db = [_RAnalyticsDatabase databaseWithConnection:connection];
    
    return db;
}

BOOL isTableExist(NSString* table, sqlite3* connection) {
    int tableCount = 0;
    NSString* query = [NSString stringWithFormat:@"SELECT count(*) FROM sqlite_master WHERE type='table' AND name='%@'", table];
    
    sqlite3_stmt* statement;
    sqlite3_prepare_v2(connection, query.UTF8String, -1, &statement, 0);
    sqlite3_step(statement);
    tableCount = sqlite3_column_int(statement, 0);
    sqlite3_reset(statement);
    sqlite3_finalize(statement);
    
    return tableCount > 0;
}

NSArray* fetchTableContents(NSString* table, sqlite3* connection) {
    NSMutableArray* result = [NSMutableArray array];
    
    NSString *query = [NSString stringWithFormat:@"select * from %@", table];
    
    sqlite3_stmt *statement;
    if (sqlite3_prepare_v2(connection, query.UTF8String, -1, &statement, 0) == SQLITE_OK) {
        int code;
        while ((code = sqlite3_step(statement)) == SQLITE_ROW) {
            const void *bytes = sqlite3_column_blob(statement, 1);
            NSUInteger length = (NSUInteger)sqlite3_column_bytes(statement, 1);
            
            [result addObject:[NSData dataWithBytes:bytes length:length]];
        }
        sqlite3_finalize(statement);
    }
    
    return result;
}

void insertBlobsIntoTable(NSArray* blobs, NSString* table, sqlite3* connection) {
    NSString *createTableQuery = [NSString stringWithFormat:@"create table if not exists %@ (id integer primary key, data blob)", table];
    sqlite3_exec(connection, createTableQuery.UTF8String, 0, 0, 0);
    
    
    NSString *insertQuery = [NSString stringWithFormat:@"insert into %@ (data) values(?)", table];
    for (NSData *blob in blobs) {
        sqlite3_stmt *statement;
        
        sqlite3_prepare_v2(connection, insertQuery.UTF8String, -1, &statement, 0);
        sqlite3_bind_blob(statement, 1, blob.bytes, (int)blob.length, 0);
        sqlite3_step(statement);
        sqlite3_clear_bindings(statement);
        sqlite3_reset(statement);
        sqlite3_finalize(statement);
    }
    
    sqlite3_exec(connection, "commit transaction", 0, 0, 0);
}

File: UnitTestApp/Tests/RAnalyticsTestHelpers/DatabaseTestUtils.swift
import Foundation
import SQLite3
@testable import RakutenAnalytics

@objc public final class DatabaseTestUtils: NSObject {
    @objc public static func openRegularConnection() -> SQlite3Pointer? {
        return openConnection(SQLITE_OPEN_READWRITE | SQLITE_OPEN_CREATE | SQLITE_OPEN_FULLMUTEX)
    }

    @objc public static func openReadonlyConnection() -> SQlite3Pointer? {
        return openConnection(SQLITE_OPEN_READONLY) // will fail if database is not created
    }

    @objc static func openConnection(_ flags: Int32) -> SQlite3Pointer? {
        var connection: SQlite3Pointer?

        // With 2 connections open we cannot use temporary databases because they are not shared.
        // Using memory databases allow sharing (cache=shared) but because of that the readonlyConnection is able to write the database.
        // https://stackoverflow.com/questions/40547077/is-it-possible-to-connect-to-an-in-memory-sqlite-db-in-read-only-mode
        assert(sqlite3_open_v2("file::memory:?cache=shared", &connection, flags, nil) == SQLITE_OK)
        if (flags & SQLITE_OPEN_READONLY) != 0 {
            assert(sqlite3_exec(connection, "PRAGMA query_only = 1", nil, nil, nil) == SQLITE_OK) // forces read-only behavior
        }

        return connection
    }

    public static func mkDatabase(connection: SQlite3Pointer) -> RAnalyticsDatabase {
        return RAnalyticsDatabase.database(connection: connection)
    }

    @objc public static func isTablePresent(_ table: String, connection: SQlite3Pointer) -> Bool {
        let query = "SELECT EXISTS(SELECT name FROM sqlite_master WHERE type='table' AND name='\(table)')"

        var statement: SQlite3Pointer?
        sqlite3_prepare_v3(connection, query, -1, 0, &statement, nil)
        sqlite3_step(statement)
        let tableCount = sqlite3_column_int(statement, 0)
        sqlite3_reset(statement)
        sqlite3_finalize(statement)

        return tableCount > 0
    }

    @objc public static func deleteTableIfExists(_ table: String, connection: SQlite3Pointer) {
        let query = "DROP TABLE IF EXISTS '\(table)'"

        var statement: SQlite3Pointer?
        sqlite3_prepare_v3(connection, query, -1, 0, &statement, nil)
        sqlite3_step(statement)
        sqlite3_reset(statement)
        sqlite3_finalize(statement)
    }

    @objc public static func fetchTableContents(_ table: String, connection: SQlite3Pointer, errorCallback: ((String) -> Void)? = nil) -> [Data] {
        var result = [Data]()
        let query = "select * from \(table)"

        var statement: SQlite3Pointer?
        
        let code = sqlite3_prepare_v3(connection, query, -1, 0, &statement, nil)
        switch code {
        case SQLITE_OK:
            while sqlite3_step(statement) == SQLITE_ROW {
                let bytes: UnsafeRawPointer = sqlite3_column_blob(statement, 1)
                let length = sqlite3_column_bytes(statement, 1)

                result.append(Data(bytes: bytes, count: Int(length)))
            }
            sqlite3_finalize(statement)
        default:
            var errorMessage = "DatabaseTestUtils: fetchTableContents() error. Unexpected sqlite code \(code)"
            if let sqliteError = sqlite3_errmsg(connection) {
                errorMessage += " - \(String(cString: sqliteError))"
            }
            errorCallback?(errorMessage)
        }

        return result
    }

    @objc public static func insert(blobs: [Data], table: String, connection: SQlite3Pointer) {
        assert(sqlite3_exec(connection, "begin exclusive transaction", nil, nil, nil) == SQLITE_OK)

        let createTableQuery = "create table if not exists \(table) (id integer primary key, data blob)"
        assert(sqlite3_exec(connection, createTableQuery, nil, nil, nil) == SQLITE_OK)

        let insertQuery = "insert into \(table) (data) values(?)"
        var statement: SQlite3Pointer?
        assert(sqlite3_prepare_v3(connection, insertQuery, -1, 0, &statement, nil) == SQLITE_OK)

        blobs.forEach { blob in
            blob.withUnsafeBytes { bytes -> Void in
                let status1 = sqlite3_bind_blob(statement, 1, bytes.baseAddress, Int32(blob.count), nil)
                let status2 = sqlite3_step(statement)
                let status3 = sqlite3_clear_bindings(statement)
                let status4 = sqlite3_reset(statement)
                assert([status1, status3, status4].allSatisfy({ $0 == SQLITE_OK }))
                assert(status2 == SQLITE_DONE)
            }
        }

        sqlite3_finalize(statement)
        assert(sqlite3_exec(connection, "commit transaction", nil, nil, nil) == SQLITE_OK)
    }

    @objc public static func databaseName(connection: SQlite3Pointer) -> String? {
        guard let dbName = sqlite3_db_filename(connection, nil) else {
            return nil
        }
        return String(cString: dbName)
    }
}
